drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -635,15 +635,27 @@ int usb_dc_ep_read_continue(uint8_t ep)
 		return -EINVAL;
 	}
 
-	bytes_received = state.setup_available ?
-			 sizeof(struct usb_setup_packet) :
-			 udc_rpi_get_ep_buffer_len(ep);
+	bool data_packet_after_setup_is_expected = false;
+	if (state.setup_available) {
+		const struct usb_setup_packet *setup = (const void *)&usb_dpram->setup_packet;
+		if ((setup->wLength != 0U) ||
+		    !usb_reqtype_is_to_host(setup)) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint exactly once, to either:
+			 * 1) receive the data in the to_device case (only if wLength is not 0)
+			 * 2) receive a 0-length status stage transfer in the to_host case
+			 */
+			  data_packet_after_setup_is_expected = true;
+		}
+	}
 
 	state.setup_available = false;
 
 	/* If no more data in the buffer, start a new read transaction. */
 	LOG_DBG("received %d offset: %d", bytes_received, ep_state->read_offset);
-	if (bytes_received == ep_state->read_offset) {
+
+	if (((ep != USB_CONTROL_EP_OUT) && (udc_rpi_get_ep_buffer_len(ep) == ep_state->read_offset)) || 
```

Comment:
Isn't `(ep != USB_CONTROL_EP_OUT) && (udc_rpi_get_ep_buffer_len(ep) == ep_state->read_offset)` always false due to `ep` guaranteed to be `USB_CONTROL_EP_OUT` at this point? Note that the code at beginning of this function will return `-EINVAL` when `!USB_EP_DIR_IS_OUT(ep)` is true.